### DISCUSSION TITLE
Read Adhar number from the QR Code ?

### DISCUSSION THREAD

**Discussion by dwishaventures** [tanmoysrt/pyaadhaar#2] [Q&A] [answered]
Hi Tanmoy, 

Can we read the Adhar Number from the QR code ? 

Thank you.

**Comment by tanmoysrt** [comment, marked as answer]
In the latest QR codes , UIDAI doesn't allow access to full aadhaar no .
You can get only the last 4 digits of an aadhar card.
But for old QR codes, you can access full aadhar no / uid .

Thanking You
Tanmoy Sarkar
Founder of Quikshope
ᐧ

**Reply by dwishaventures** [reply]
How can I contact you , please email me dwishaventures@gmail . Thank you.

**Reply by tanmoysrt** [reply]
I have mailed you. Check

**Reply by jl-max** [reply]
Hi Tanmoy,

Can you tell me where to find the **key names** contained in a new aadhaar QRcode please ? I coudn't find any specific documents in https://uidai.gov.in/

**key names** in old aadhaar QRcode are shown in the red square:
![000](https://user-images.githubusercontent.com/64684153/176650072-ee0adc32-de10-44ac-aee0-922a7751aed4.png)

Thank you.

**Reply by tanmoysrt** [reply]
If you want to use pyaadhaar library, You need to use
```
from pyaadhaar.decode import AadhaarSecureQr

obj  = AadhaarSecureQr(<:Data got from QR:>)
print(obj.decodeddata())
```

If you want to understand how to extract data from the encoded string of latest secure qr, please go through this PDF and the implementation of AadhaarSecureQr function

[User_manulal_QR_Code_15032019.pdf](https://github.com/Tanmoy741127/pyaadhaar/files/9019080/User_manulal_QR_Code_15032019.pdf)

If you have any doubt, feel free to reach me .

**Reply by jl-max** [reply]
Thanks a lot for your generous help! The PDF does help.

**Reply by jl-max** [reply]
Which data format should I send in as <:Data got from QR:>?

**Comment by tanmoysrt** [comment]
Scan the qr, you will get data in binary format consisting of numbers only.
Pass that data as a string.

**Reply by jl-max** [reply]
You mean pass the data in string ? but I got an error like this: 
![error](https://user-images.githubusercontent.com/64684153/176860296-cd848857-062b-4ccd-8763-e29980a1b0a1.png)

**Reply by sahiljadhav123** [reply]
same error

**Reply by tanmoysrt** [reply]
@jl-max @sahiljadhav123 

If you have a string, you need to encode it before passing it to the constructor
```python
z = "Hello World"
encoded_z = z.encode()
```

pass the encoded_z to the constructor

I think it will fix your issue

**Reply by sahiljadhav123** [reply]
self.data['adhaar_last_4_digit'] = self.data['referenceid'][0:4]
---> 40 self.data['adhaar_last_digit'] = self.data['referenceid'][3]
     42 if self.data['email_mobile_status'] == "0":
     43     self.data['email'] = "no"

IndexError: string index out of range


error in   self.data['referenceid'][3]

**Reply by sahiljadhav123** [reply]
masking adhar Qr code decode ? is possible

**Reply by tanmoysrt** [reply]
@sahiljadhav123 I haven't tested actually, I need to check 
There were not any specific docs at all for masked aadhar card.
I will check and update you

**Reply by sahiljadhav123** [reply]
One more issue-  invalid literal for int () base 10 :

**Reply by tanmoysrt** [reply]
I will check the issues and will try to resolve them and release a new version. For the past few months, can't get time to release the new version

**Comment by tanmoysrt** [comment]
OK I will check and let you know

**Comment by tanmoysrt** [comment]
Sorry, You need to pass the raw data directly to that function. The long
integer that you got from QR code. No need to convert to string explicitly.
ᐧ

**Comment by tanmoysrt** [comment]
Use pyzbar

**Comment by samruddhisk** [comment]
Describe the bug
When I try to retrieve the QR data from new aadhaar QR code it throws error as 'string out of index'. Whereas if i try with old aadhaar QR code data, I am getting the correct expected output.

Expected behavior
For the new QR Code also i must get the proper result without error. I have updated my pyaadhar package to 2.0.2 and also my python version is 3.12

Additional context
But as per your new release you have mentioned as, it supports the old and new both aadhaar QR codes.
Could you please help me in sorting out this.

**Reply by tanmoysrt** [reply]
I am not able to find current specs of secure aadhar QR.
If specs has been changed, then it's a issue with parsing.

If you found the specs sheet, send here also

**Reply by samruddhisk** [reply]
Hey Tanmay,
Thanks for your response.
Even we are using the same user manual which you've referred to. We too didn't find any updated document for the same.
Anyways I have tried a different approach to handle QR data based on the different versions. I have enhanced the decode.py file's AadhaarSecureQR  function and it is able to extract the data for both old and new aadhaar QR codes. I'm sharing you the file regarding the same, PFA. But as we are using pyaadhaar library, the decode.py file is under that, so changing the function code in library package is not a feasible solution when it comes to deploying to other servers. Anyways my updated works in my local system server. But the problem arises in other servers. Please go through my code once and see if that can be helpful for you too to integrate with your library function. Also I'm attaching a document regarding the changes I made.
[myDecode(py)code.txt](https://github.com/user-attachments/files/16343739/myDecode.py.code.txt)
[AadhaarSecureQr Code Comparison.docx](https://github.com/user-attachments/files/16343740/AadhaarSecureQr.Code.Comparison.docx)

**Reply by tanmoysrt** [reply]
Raise a PR
I will merge and release a new version

**Reply by samruddhisk** [reply]
Yeah sure

**Reply by samruddhisk** [reply]
I have raised a PR. you can go through it.
